- en: Specifying the Exceptions Thrown by a Method
  id: totrans-0
  prefs:
  - PREF_H1
  type: TYPE_NORMAL
  zh: 指定方法抛出的异常
- en: 原文：[https://docs.oracle.com/javase/tutorial/essential/exceptions/declaring.html](https://docs.oracle.com/javase/tutorial/essential/exceptions/declaring.html)
  id: totrans-1
  prefs:
  - PREF_BQ
  type: TYPE_NORMAL
  zh: 原文：[https://docs.oracle.com/javase/tutorial/essential/exceptions/declaring.html](https://docs.oracle.com/javase/tutorial/essential/exceptions/declaring.html)
- en: The previous section showed how to write an exception handler for the `writeList`
    method in the `ListOfNumbers` class. Sometimes, it's appropriate for code to catch
    exceptions that can occur within it. In other cases, however, it's better to let
    a method further up the call stack handle the exception. For example, if you were
    providing the `ListOfNumbers` class as part of a package of classes, you probably
    couldn't anticipate the needs of all the users of your package. In this case,
    it's better to *not* catch the exception and to allow a method further up the
    call stack to handle it.
  id: totrans-2
  prefs: []
  type: TYPE_NORMAL
  zh: 前一节展示了如何为`ListOfNumbers`类中的`writeList`方法编写异常处理程序。有时，代码捕获可能发生的异常是合适的。然而，在其他情况下，最好让调用堆栈中更高层的方法处理异常。例如，如果你将`ListOfNumbers`类作为一个类包的一部分提供，你可能无法预料到包的所有用户的需求。在这种情况下，最好*不要*捕获异常，而是让调用堆栈中更高层的方法来处理它。
- en: If the `writeList` method doesn't catch the checked exceptions that can occur
    within it, the `writeList` method must specify that it can throw these exceptions.
    Let's modify the original `writeList` method to specify the exceptions it can
    throw instead of catching them. To remind you, here's the original version of
    the `writeList` method that won't compile.
  id: totrans-3
  prefs: []
  type: TYPE_NORMAL
  zh: 如果`writeList`方法不捕获其中可能发生的已检查异常，那么`writeList`方法必须指定它可以抛出这些异常。让我们修改原始的`writeList`方法，以指定它可以抛出的异常，而不是捕获它们。为了提醒你，这里是原始版本的`writeList`方法，它不会编译通过。
- en: '[PRE0]'
  id: totrans-4
  prefs: []
  type: TYPE_PRE
  zh: '[PRE0]'
- en: To specify that `writeList` can throw two exceptions, add a `throws` clause
    to the method declaration for the `writeList` method. The `throws` clause comprises
    the `throws` keyword followed by a comma-separated list of all the exceptions
    thrown by that method. The clause goes after the method name and argument list
    and before the brace that defines the scope of the method; here's an example.
  id: totrans-5
  prefs: []
  type: TYPE_NORMAL
  zh: 要指定`writeList`可能会抛出两个异常，需要在`writeList`方法的方法声明中添加一个`throws`子句。`throws`子句包括`throws`关键字，后面跟着一个逗号分隔的由该方法抛出的所有异常的列表。该子句放在方法名和参数列表之后，方法范围定义的大括号之前；以下是一个示例。
- en: '[PRE1]'
  id: totrans-6
  prefs: []
  type: TYPE_PRE
  zh: '[PRE1]'
- en: Remember that `IndexOutOfBoundsException` is an unchecked exception; including
    it in the `throws` clause is not mandatory. You could just write the following.
  id: totrans-7
  prefs: []
  type: TYPE_NORMAL
  zh: 请记住`IndexOutOfBoundsException`是一个未经检查的异常；在`throws`子句中包含它并不是强制性的。你可以只写如下内容。
- en: '[PRE2]'
  id: totrans-8
  prefs: []
  type: TYPE_PRE
  zh: '[PRE2]'
